packages/editor/src/extensions/key-map/move-node.ts

```diff
@@ -0,0 +1,215 @@
+import { OutlineList } from "../outline-list/outline-list.js";
+import { Table } from "../table/table.js";
+import { TaskListNode } from "../task-list/task-list.js";
+
+type ResolvedNode = {
+  pos: number;
+  start: number;
+  depth: number;
+  node: Node;
+};
+
+function mapChildren<T>(
+  node: Node | Fragment,
+  callback: (child: Node, index: number, frag: Fragment) => T
+): T[] {
+  const array = [];
+  for (let i = 0; i < node.childCount; i++) {
+    array.push(
+      callback(node.child(i), i, node instanceof Fragment ? node : node.content)
+    );
+  }
+  return array;
+}
+
+function resolveNode(editor: Editor) {
+  const { state } = editor;
+  const { $from } = state.selection;
+
+  let targetType = $from.node().type;
+  let currentResolved = findParentNodeOfType(targetType)(state.selection);
+
+  if (isListActive(editor)) {
+    const currentNode = $from.node();
+    const parentNode = $from.node($from.depth - 1);
+    const isFirstParagraph =
+      currentNode.type.name === "paragraph" &&
+      parentNode.firstChild === currentNode;
+
+    // move the entire list item
+    if (isFirstParagraph) {
+      targetType = $from.node($from.depth - 1).type;
+      if (
+        targetType.name === Callout.name ||
+        targetType.name === Blockquote.name
+      ) {
+        targetType = $from.node($from.depth - 2).type;
+      }
+    }
+
+    currentResolved = findParentNodeOfType(targetType)(state.selection);
+  }
+
+  if (
+    findParentNodeClosestToPos($from, (node) => node.type.name === Callout.name)
+  ) {
+    const currentNode = $from.node();
+    const parentNode = $from.node($from.depth - 1);
+    const isFirstHeading =
+      currentNode.type.name === "heading" &&
+      parentNode.firstChild === currentNode;
+
+    // move the entire callout
+    if (isFirstHeading) {
+      targetType = $from.node($from.depth - 1).type;
+    }
+
+    currentResolved = findParentNodeOfType(targetType)(state.selection);
+  }
+
+  return currentResolved;
+}
+
+function resolveParentNode(editor: Editor) {
+  const { state } = editor;
+  const { $from } = state.selection;
+
+  const validParents = [
+    Callout.name,
+    Table.name,
+    BulletList.name,
+    OrderedList.name,
+    TaskListNode.name,
+    CheckList.name,
+    OutlineList.name,
+    Blockquote.name
+  ];
+
+  const parent = findParentNodeClosestToPos($from, (node) =>
+    validParents.includes(node.type.name)
+  );
+
+  if (!parent) return undefined;
+
+  const targetType = parent.node.type;
+  return findParentNodeOfType(targetType)(state.selection);
+}
+
+function swapNodeWithSibling(
+  editor: Editor,
+  resolvedNode: ResolvedNode,
+  isDown: boolean
+): boolean {
+  const { state } = editor;
+  const { $from } = state.selection;
+
+  const parentDepth = resolvedNode.depth - 1;
+  const parent = $from.node(parentDepth);
+  const parentPos = $from.start(parentDepth);
+
+  let arr = mapChildren(parent, (node) => node);
+  let index = arr.indexOf(resolvedNode.node);
```

Comment:
Why do you need to map all the children into an array in order to find the index of the child?

Reply:
This `arr` is used for swapping, so we need to find the index and the corresponding swapping index